File: workers/python/activities.py
import asyncio
import os

from google.protobuf.duration_pb2 import Duration
from temporalio import activity
from temporalio.client import Client
from temporalio.exceptions import ApplicationError

from client_action_executor import ClientActionExecutor


@activity.defn(name="noop")
async def noop_activity():
    return


@activity.defn(name="delay")
async def delay_activity(delay_for: Duration):
    await asyncio.sleep(delay_for.ToSeconds())


@activity.defn(name="payload")
async def payload_activity(input_data: bytes, bytes_to_return: int) -> bytes:
    return os.urandom(bytes_to_return)


@activity.defn(name="retryable_error")
async def retryable_error_activity(config):
    """Activity that throws retryable errors for N attempts, then succeeds."""
    info = activity.info()
    if info.attempt <= config.fail_attempts:
        raise ApplicationError("retryable error", "RetryableError", non_retryable=False)


@activity.defn(name="timeout")
async def timeout_activity(config):
    """Activity that runs too long for N attempts (causing timeout), then completes quickly."""
    info = activity.info()
    duration = config.success_duration
    if info.attempt <= config.fail_attempts:
        # Failure case: run failure duration (exceeds activity timeout)
        duration = config.failure_duration

    # Sleep for failure/success timeout duration.
    # In failure case, this will throw a cancellation error.
    await delay_activity(duration)


@activity.defn(name="heartbeat")
async def heartbeat_activity(config):
    """Activity that skips heartbeats for N attempts (causing heartbeat timeout), then sends them."""
    info = activity.info()
    should_send_heartbeats = info.attempt > config.fail_attempts
    if not should_send_heartbeats:
        # Failure case: run failure duration (exceeds heartbeat timeout)
        await delay_activity(config.failure_duration)
        return

    duration = (
        config.success_duration.seconds + config.success_duration.nanos / 1_000_000_000
    )
    interval = (
        config.heartbeat_interval.seconds
        + config.heartbeat_interval.nanos / 1_000_000_000
    )
    if interval <= 0:
        await asyncio.sleep(duration)
        activity.heartbeat()
        return

    await run_with_heartbeats(duration, interval)


async def run_with_heartbeats(duration, interval):
    activity.heartbeat()
    remaining = duration
    while remaining > 0:
        sleep_for = min(interval, remaining)
        await asyncio.sleep(sleep_for)
        remaining -= sleep_for
        if remaining > 0:
            activity.heartbeat()


def create_client_activity(client: Client, err_on_unimplemented: bool):
    @activity.defn(name="client")
    async def client_activity(client_activity_proto):
        activity_info = activity.info()
        workflow_id = activity_info.workflow_id
        executor = ClientActionExecutor(
            client, workflow_id, activity_info.task_queue, err_on_unimplemented
        )
        await executor.execute_client_sequence(client_activity_proto.client_sequence)

    return client_activity

File: workers/ruby/activities.rb
require 'temporalio/activity'
require 'temporalio/activity/definition'
require_relative 'client_action_executor'

class NoopActivity < Temporalio::Activity::Definition
  activity_name 'noop'

  def execute
    nil
  end
end

class DelayActivity < Temporalio::Activity::Definition
  activity_name 'delay'

  def execute(delay_for)
    sleep(delay_for.seconds + (delay_for.nanos / 1_000_000_000.0))
  end
end

class PayloadActivity < Temporalio::Activity::Definition
  activity_name 'payload'

  def execute(_input_data, bytes_to_return)
    Random.bytes(bytes_to_return)
  end
end

class RetryableErrorActivity < Temporalio::Activity::Definition
  activity_name 'retryable_error'

  def execute(config)
    info = Temporalio::Activity::Context.current.info
    return unless info.attempt <= config.fail_attempts

    raise Temporalio::Error::ApplicationError.new('retryable error', type: 'RetryableError')
  end
end

class TimeoutActivity < Temporalio::Activity::Definition
  activity_name 'timeout'

  def execute(config)
    info = Temporalio::Activity::Context.current.info
    duration = config.success_duration
    duration = config.failure_duration if info.attempt <= config.fail_attempts
    sleep(duration.seconds + (duration.nanos / 1_000_000_000.0))
  end
end

class HeartbeatActivity < Temporalio::Activity::Definition
  activity_name 'heartbeat'

  def execute(config)
    info = Temporalio::Activity::Context.current.info
    should_send_heartbeats = info.attempt > config.fail_attempts
    duration = should_send_heartbeats ? config.success_duration : config.failure_duration
    DelayActivity.new.execute(duration)
    Temporalio::Activity::Context.current.heartbeat if should_send_heartbeats
  end
end

class ClientActivity < Temporalio::Activity::Definition
  activity_name 'client'

  def initialize(client, err_on_unimplemented: false)
    super()
    @client = client
    @err_on_unimplemented = err_on_unimplemented
  end

  def execute(client_activity_proto)
    info = Temporalio::Activity::Context.current.info
    executor = ClientActionExecutor.new(
      @client,
      info.workflow_id,
      info.task_queue,
      err_on_unimplemented: @err_on_unimplemented
    )
    executor.execute_client_sequence(client_activity_proto.client_sequence)
  end
end
